### ISSUE TITLE
"node server": endless loop copying files

### ISSUE THREAD

**Issue by lll000111** [wkh237/react-native-fetch-blob-dev#1]
EDITED:

Since my problems have changed throughout the experiments - to have a starting point I tried to install everything as-is and run it but that leads to many errors.

BUILD error description removed, using my own code and updating the image picker as described in the next comment got rid of that.

When I try to run `node server` I get

```
fs.js:652
  return binding.open(pathModule._makeLong(path), stringToFlags(flags), mode);
                 ^

Error: EPERM: operation not permitted, open 'D:\Documents\Projects\react-native-fetch-blob-dev\src\.git'
    at Object.fs.openSync (fs.js:652:18)
    at Object.fs.writeFileSync (fs.js:1299:33)
    at D:\Documents\Projects\react-native-fetch-blob-dev\test-server\server.js:300:12
    at D:\Documents\Projects\react-native-fetch-blob-dev\node_modules\mkdirp\index.js:48:26
    at FSReqWrap.oncomplete (fs.js:153:5)
```

This is an impossible error, I'm doing everything as the same user and I have full permissions for that file:

```
$ ls -l /d/Documents/Projects/react-native-fetch-blob-dev/src/.git
-rw-r--r-- 1 mha 197609 28 Aug 14 12:57 /d/Documents/Projects/react-native-fetch-blob-dev/src/.git

mha@mha-MAINPC MINGW64 /D/Documents/Projects/react-native-fetch-blob-dev/test-server (master)
$ cat /d/Documents/Projects/react-native-fetch-blob-dev/src/.git
gitdir: ../.git/modules/src
```

When I put a `console.log(111111, targetPath, path)` in front of the error-causing line I get

> `'..\\src\\.git' '..\\src\\.git'`

But *all* calls are of this pattern (not that I understand why the code copies files onto themselves?) but only that one causes that message. For example, the first one with the two variable values `'..\\test\\.flowconfig' '..\\test\\.flowconfig'` worked just fine...

*EDIT:* When I use my own repo this same problem appears at a different point:

```
file changed ..\src\.git\hooks\update.sample
fs.js:652
  return binding.open(pathModule._makeLong(path), stringToFlags(flags), mode);
                 ^

Error: EPERM: operation not permitted, open 'D:\Documents\Projects\react-native-fetch-blob-dev\src\.git\objects\pack\pack-5a17f2a83dd647dbf3bacaa904c799a4ec54f500.idx'
```

<br>

I have no idea what to do. I'm running out of ideas, options and time. I don't think there should be _much_ wrong with my patch(es) over at https://github.com/lll000111/react-native-fetch-blob/tree/0.10.9, but I just can't run those tests.

**By the way,** what is missing from the instructions is how to use my own folk of the repo. I can just manually clone it to `./src/`, but I don't want to learn how to manipulate the git submodule stuff you already setup. I guess I can just ignore it apart from the initial install, I think don't need git any more for this test-dev repo?

## Q: The code in Android studio

When I open the (thus far unchanged) `./src/android/` code in Android studio I already get numerous errors: https://gist.github.com/lll000111/423b858d707daf5eba5212ca5e1f6baa

Shouldn't the project be error-free when I use that IDE?

If not, what do I have to do to make it error-free? Shouldn't I be able to build it in Android Studio?

If not, what IDE and environment are <i>you</i> using for development?

<br>
Thanks!

**Comment by lll000111** [comment]
@wkh237 On the bright side, I've done a few simple tests that include the new error code (err.code) and the new `hash` function and that at least works.

(Note: the initial comment was about a plain RNFB-dev package and whatever version of RNFB it installs. Below is about my own test project directory, based on a basic `react-native init TestProject`, using my own repo and it 0.10.9 branch (https://github.com/lll000111/react-native-fetch-blob/tree/0.10.9).

VERY simple and quick test:

```javascript
RNFetchBlob.fs.mkdir('/storage/emulated/0/ONE')
.catch(err => {
    console.log('MKDIR', err.code, err.name, err.message);
    if (!err.message.includes('already exists')) {
        throw err;
    }
})
.then(() => RNFetchBlob.fs.exists('/storage/emulated/0/ONE/DEMO-FILE.txt'))
.then(exists => {
    console.log('EXISTS', exists);
    if (exists) {
        return RNFetchBlob.fs.unlink('/storage/emulated/0/ONE/DEMO-FILE.txt');
    }
})
.then(() => RNFetchBlob.fs.createFile('/storage/emulated/0/ONE/DEMO-FILE.txt', 'DEMOooooo 1111 CONTENT'))
.then(() => RNFetchBlob.fs.createFile('/storage/emulated/0/ONE/DEMO-FILE.txt', 'DEMOooooo 1111 CONTENT'))
.catch(err => {
    console.log('CREATE ERROR:', err.code, err.name, err.message);
})
.then(() => RNFetchBlob.fs.readFile('/storage/emulated/0/ONE/NO-SUCH-FILE', 'utf8'))
.catch(err => {
    console.log('CREATE ERROR:', err.code, err.name, err.message);
})
.then(() => RNFetchBlob.fs.readFile('/storage/emulated/0/ONE/DEMO-FILE.txt', 'utf8'))
.then(result => console.log(1, result))
.then(() => RNFetchBlob.fs.writeFile('/storage/emulated/0/ONE/DEMO-FILE.txt', 'DEMOooooo 222 CONTENT'))
.then(() => RNFetchBlob.fs.readFile('/storage/emulated/0/ONE/DEMO-FILE.txt', 'utf8'))
.then(result => console.log(2, result))
.then(() => RNFetchBlob.fs.hash('/storage/emulated/0/ONE/DEMO-FILE.txt', 'sha256'))
.then(result => console.log(3, result))
.then(() => responseJson)
```

The result of that tiny test:

```
'MKDIR', 'EEXIST', 'Error', 'Folder \'/storage/emulated/0/ONE\' already exists'
'EXISTS', true
'CREATE ERROR:', 'EEXIST', 'Error', 'File `/storage/emulated/0/ONE/DEMO-FILE.txt` already exists'
'CREATE ERROR:', 'ENOENT', 'Error', 'No such file or directory \'/storage/emulated/0/ONE/NO-SUCH-FILE\'; /storage/emulated/0/ONE/NO-SUCH-FILE: open failed: ENOENT (No such file or directory)'
1, 'DEMOooooo 1111 CONTENT'
2, 'DEMOooooo 222 CONTENT'
3, 'be04da7bae1145ce4f5d2df9720fafb6e01182ff4e04f801413d805eef1b27d5'
```

<br>

Now, if only I could get the tests to run too...

**Comment by lll000111** [comment]
I think the image-pickernpm module needs to be updated to version 0.26.4 in package.json: https://github.com/react-community/react-native-image-picker/pull/653

Without it there is an error:

```
D:\Documents\Projects\react-native-fetch-blob-dev\RNFetchBlobTest\node_modules\react-native-image-picker\android\src\main\java\com\imagepicker\ImagePickerPackage.java:34: error: method does not override or implement a method from a supertype
  @Override
  ^
1 error
:react-native-image-picker:compileReleaseJavaWithJavac FAILED
```

<br>

The app starts after I update this, but I only see the default "hello world" app screen. I guess  need that server to run...?

**Comment by lll000111** [comment]
When I move the `.git` directory away I don't get errors any more - but the server is in an endless `file changed` loop, copying the same files over and over.

**Comment by lll000111** [comment]
When I add a statement `console.log(targetPath, path);` to the `on('add',..)` event handler of the watcher (in line 296) start the server I get this:

```
$ node server.js
SSL test server running at port  8124
test server running at port  8123
..\src\.gitignore ..\src\.gitignore
..\src\CONTRIBUTING.md ..\src\CONTRIBUTING.md
..\src\CONTRIBUTORS.md ..\src\CONTRIBUTORS.md
..\src\README.md ..\src\README.md
..\src\LICENSE ..\src\LICENSE
..\src\android.js ..\src\android.js
..\src\fs.js ..\src\fs.js
..\src\index.js ..\src\index.js
..\src\ios.js ..\src\ios.js
..\src\json-stream.js ..\src\json-stream.js
..\src\package.json ..\src\package.json
..\src\react-native-fetch-blob.podspec ..\src\react-native-fetch-blob.podspec
..\src\types.js ..\src\types.js
...(etc.)
```

Source and destination are the same.

**Comment by lll000111** [comment]
**SLASHES! The problem are "/" vs. "\".** I'm running on Windows. When I add

    console.log(targetPath, path, source, dest);

In line 297, in the `on('add',...)` event handler of the watcher, I get this result:

```
$ node server.js
SSL test server running at port  8124
test server running at port  8123
..\test\.flowconfig ..\test\.flowconfig ../test/ ../RNFetchBlobTest/
..\test\benchmark.js ..\test\benchmark.js ../test/ ../RNFetchBlobTest/
..\test\index.android.js ..\test\index.android.js ../test/ ../RNFetchBlobTest/
..\test\index.ios.js ..\test\index.ios.js ../test/ ../RNFetchBlobTest/
..\test\index.test.js ..\test\index.test.js ../test/ ../RNFetchBlobTest/
..\test\nedb.js ..\test\nedb.js ../test/ ../RNFetchBlobTest/
...(etc.)
```

So the `String(path).replace(source, dest)` in line 296 does not do anything.

**Comment by lll000111** [comment]
And now I end up with

![image](https://user-images.githubusercontent.com/25476357/29284214-1293255a-812a-11e7-98c9-614fb5e7928e.png)

**Comment by lll000111** [comment]
I got the tests to run, but it stops with two unhandled promise rejections and 

```
Possible Unhandled Promise Rejection (id: 1):
Error: Failed to connect to localhost/127.0.0.1:8123
http://10.0.2.2:8081/index.android.bundle?platform=android&dev=true&hot=false&minify=false:60368:32
__invokeCallback@http://10.0.2.2:8081/index.android.bundle?platform=android&dev=true&hot=false&minify=false:4887:21
http://10.0.2.2:8081/index.android.bundle?platform=android&dev=true&hot=false&minify=false:4711:32
__guard@http://10.0.2.2:8081/index.android.bundle?platform=android&dev=true&hot=false&minify=false:4822:11
invokeCallbackAndReturnFlushedQueue@http://10.0.2.2:8081/index.android.bundle?platform=android&dev=true&hot=false&minify=false:4710:19
invokeCallbackAndReturnFlushedQueue@[native code]
```

![image](https://user-images.githubusercontent.com/25476357/29284403-e8f1960e-812a-11e7-9b61-8ae45b8201e0.png)



The screen stops at this state:

![image](https://user-images.githubusercontent.com/25476357/29284464-2be77992-812b-11e7-915f-6c34947c3603.png)


<br>

<br>

On repeat I don't get it to communicate (and run) at all. The app-GUI is up, the server is up - nothing happens, one by one the items on the screen turn to red timeouts.

**Comment by lll000111** [comment]
I'll close it for now until I figure out what's going on. Apparently I don't have network enabled in the emulator.

**Comment by wkh237** [comment]
@lll000111 , thanks for the great information. However it's working fine on my end (Mac). Never tried it on windows or linux. Perhaps that's something related to how `watch` is working on different platform.

**Comment by lll000111** [comment]
I'll submit a PR for this repo with additonal tests for "fs", like the new hash function, and the new error structure (with a `code` property as implemented in the react-native promise bridge code). I added `path = path.replace(/\\/g, '/');` to all three "chokidar" event handler functions, which should have no effect on platforms with only "/" but fix the problem on Windows. Yes, definitely an issue in the "chokidar" package. I also added `.git/` to the "chokidar" ignore pattern so that one can use ones own repo directly from github and do some development.

My tests now run and I'm at 100% - of the stuff that I'm testing, which is mostly "fs", where almost all my editing took place.

I'll prepare the PR later today or tomorrow - I have more trouble with git and creating a PR than with the actual code... if I just create a new PR based on the same branch as before the previous commits are included. "Just create a new branch" - well, I tried many combinations of that and I **hate** git. If I used mere text patches this would be no problem at all, git makes it MUCH more complicated,

<br>

PS: The network problem went away when I switched the `localhost:8123` against `10.0.2.2:8123` (same for https). I'm using the emulator, where the special IP address 10.0.2.2 inside the emulator points to the host's 127.0.0.1 - the emulator has its own encapsulated network environment and no direct connection to the outside.
